app/src/androidTest/java/ServiceTests/GpsSdkEndToEnd.java

```diff
@@ -0,0 +1,124 @@
+                "Successfully validated expected database state ");
+
+
+        // --------------------- Set and Validate Second Location ---------------------
+        loc2 = new Location(UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue)
+                , UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue), 3);
+        AssertUtils.assertTrueV(
+                DBHelper.setLocationAndValidateDB(loc2, mainApplication),
+                "Database state does not match with the desired state",
+                "Successfully validated expected database state ");
+
+
+        // --------------------- Set and Validate Third Location ---------------------
+        loc3 = new Location(UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue)
+                , UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue), 3);
+        AssertUtils.assertTrueV(
+                DBHelper.setLocationAndValidateDB(loc3, mainApplication),
+                "Database state does not match with the desired state",
+                "Successfully validated expected database state ");
+
+
+        // --------------------- Set and Validate Fourth Location ---------------------
+        loc4 = new Location(UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue)
+                , UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue), 3);
+        AssertUtils.assertTrueV(
+                DBHelper.setLocationAndValidateDB(loc4, mainApplication),
+                "Database state does not match with the desired state",
+                "Successfully validated expected database state ");
+
+
+        // --------------------- Set and Validate Fifth Location ---------------------
+        loc5 = new Location(UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue)
+                , UiUtils.randomGenerator(TestConstants.minValue, TestConstants.maxValue), 3);
+        AssertUtils.assertTrueV(
+                DBHelper.setLocationAndValidateDB(loc5, mainApplication),
+                "Database state does not match with the desired state",
+                "Successfully validated expected database state ");
+
+
+        // Dispatch Success Response and Validate Database
+        UiUtils.safeSleep(15);
+        edgeCaseResponses.put("/sendGps", TestConstants.RESPONSE_TYPE.SUCCESS);
+        UiUtils.safeSleep(5);
+        mockLocationList.remove(0);
+        mockLocationList.remove(0);
+        mockLocationList.remove(0);
+        AssertUtils.assertTrueV(DBHelper.validateLocationsDataInDatabase(mockLocationList, mainApplication),
+                "Database state does not match with the desired state",
+                "Successfully validated expected database state ");
+
+
+        UiUtils.safeSleep(10);
+        // There were total 5 locations we had set using mockLocation .
+        // 3 were Dispatched in last call
```

Comment:
We want this to happen through network failure. To simulate a flaky network on a bus. We need to:
- throttle 
- simulate no network response 
- and delay body

```
        MockResponse demoDelay_InBody =  new MockResponse().setResponseCode(200)
                    .setBody(apiResponse).setBodyDelay(5000, TimeUnit.MILLISECONDS);

            MockResponse demoThrottlingNetwork_ForAllResponse =  new MockResponse().setResponseCode(200)
                    .setBody(apiResponse).throttleBody(200, 8000, TimeUnit.MILLISECONDS); 
```

Explore more on this here: https://square.github.io/okhttp/3.x/mockwebserver/index.html?okhttp3/mockwebserver/MockResponse.html

Reply:
This is not for adding delay. Delay is happening in NetworkManager
![image](https://user-images.githubusercontent.com/38068550/78874230-afee3480-7a69-11ea-83a4-34cfecf2ea94.png)


This 10 seconds sleep exists because : 
  // Assuming tightest case that last sync request was made just after validating state of database in previous statement.
        // Therefore wait for another call to happen. 

I will  add comments over each sleep I have added in all tests to provide better clarity.